File: lua/atlassian/editor/ir.lua
-- Projection IR — the editor-agnostic contract between the (pure) renderer and
-- the editor-specific applier.
--
-- `render.build_ir(adf)` produces a ProjectionIR: plain data, no editor calls.
-- `render.apply_ir(buf, ir)` renders it to Neovim extmarks via to_extmark(). A
-- future VSCode or Sublime client translates the same IR to its own decoration
-- API instead.
--
-- ProjectionIR = {
--   lines        : string[]                  -- the clean buffer text
--   marks        : IRMark[]                  -- neutral decoration descriptors
--   spans        : RenderSpan[]              -- ADF path ↔ text-range map
--   line_to_node : table<number, table>      -- line → ADF node (editor-side)
--   image_refs?, link_refs? : table          -- interaction metadata (editor-side)
-- }
--
-- IRMark is a NEUTRAL descriptor — it no longer carries raw extmark opts:
--   { line, col, kind = IRDecorationKind, ...fields }
--     highlight    : { end_col, hl }
--     inline_text  : { chunks = {{text, hl?}, ...}, pos }
--     eol_text     : { chunks, align = "right_align"|"eol" }
--     sign         : { text, hl }
--     virt_lines   : { lines = {{{text, hl?}, ...}, ...}, above }
-- `hl` is a semantic style *token* (e.g. "heading.2", "strong", "diff.add"),
-- not a Neovim highlight group — see atlassian.editor.theme. Each applier maps
-- tokens to its own theme; the Neovim applier resolves them back to highlight
-- groups. `chunks`/`lines` use {text, styleToken} tuples.
--
-- The build is now editor-agnostic data end to end; the only Neovim runtime
-- dependency is display width (see atlassian.width), routed through a swappable
-- module so a non-Neovim host can provide its own.
local theme = require("atlassian.editor.theme")

local M = {}

-- Map the style token in every {text, style} chunk via `fn`, returning new chunks.
---@param chunks table[]|nil
---@param fn fun(style: string|nil): string|nil
---@return table[]|nil
local function map_chunk_styles(chunks, fn)
	if not chunks then return nil end
	local out = {}
	for i, c in ipairs(chunks) do
		out[i] = { c[1], fn(c[2]) }
	end
	return out
end

-- Map the style token in every chunk of every virtual line via `fn`.
---@param vlines table[]|nil
---@param fn fun(style: string|nil): string|nil
---@return table[]|nil
local function map_vline_styles(vlines, fn)
	if not vlines then return nil end
	local out = {}
	for i, line in ipairs(vlines) do
		out[i] = map_chunk_styles(line, fn)
	end
	return out
end

---@alias IRDecorationKind
---| "highlight"    # colored text range over [col, end_col)
---| "inline_text"  # virtual text inserted inline, shifting real text
---| "eol_text"     # virtual text at end of line / right-aligned
---| "virt_lines"   # whole virtual lines above/below (borders, tables)
---| "sign"         # sign-column glyph
---| "unknown"

--- Convert a Neovim-dialect mark ({line, col, opts}) to a neutral descriptor.
--- Called once at the build boundary so the IR a consumer sees is editor-neutral.
---@param mark table { line, col, opts }
---@return table neutral descriptor { line, col, kind, ... }
function M.to_neutral(mark)
	local o = mark.opts or {}
	local d = { line = mark.line, col = mark.col }
	if o.virt_lines then
		d.kind = "virt_lines"
		d.lines = map_vline_styles(o.virt_lines, theme.to_token)
		d.above = o.virt_lines_above or false
	elseif o.sign_text then
		d.kind = "sign"
		d.text = o.sign_text
		d.hl = theme.to_token(o.sign_hl_group)
	elseif o.virt_text then
		local pos = o.virt_text_pos
		if pos == "right_align" or pos == "eol" then
			d.kind = "eol_text"
			d.chunks = map_chunk_styles(o.virt_text, theme.to_token)
			d.align = pos
		else
			d.kind = "inline_text"
			d.chunks = map_chunk_styles(o.virt_text, theme.to_token)
			d.pos = pos or "inline"
		end
	elseif o.hl_group then
		d.kind = "highlight"
		d.hl = theme.to_token(o.hl_group)
		d.end_col = o.end_col
	else
		d.kind = "unknown"
		d.opts = o
	end
	return d
end

--- Convert a neutral descriptor back to the Neovim extmark form ({line,col,opts}).
--- This is the Neovim applier's translation step; other editors write their own.
---@param d table neutral descriptor
---@return table { line, col, opts }
function M.to_extmark(d)
	local opts
	local k = d.kind
	if k == "virt_lines" then
		opts = { virt_lines = map_vline_styles(d.lines, theme.to_nvim), virt_lines_above = d.above or nil }
	elseif k == "sign" then
		opts = { sign_text = d.text, sign_hl_group = theme.to_nvim(d.hl) }
	elseif k == "inline_text" then
		opts = { virt_text = map_chunk_styles(d.chunks, theme.to_nvim), virt_text_pos = d.pos or "inline" }
	elseif k == "eol_text" then
		opts = { virt_text = map_chunk_styles(d.chunks, theme.to_nvim), virt_text_pos = d.align }
	elseif k == "highlight" then
		opts = { hl_group = theme.to_nvim(d.hl), end_col = d.end_col }
	else
		opts = d.opts or {}
	end
	return { line = d.line, col = d.col, opts = opts }
end

--- Neutralize every mark in an IR result in place, returning the result.
---@param result table ProjectionIR with dialect marks
---@return table result with neutral marks
function M.neutralize(result)
	local marks = result.marks or {}
	for i = 1, #marks do
		marks[i] = M.to_neutral(marks[i])
	end
	return result
end

--- The portable, wire-serializable subset of an IR. Drops editor-side
--- interaction metadata (line_to_node, image_refs, link_refs) that references
--- live ADF nodes; keeps lines + neutral marks + spans (path ↔ range mapping).
---@param result table ProjectionIR (neutral marks)
---@return table { lines, marks, spans }
function M.to_wire(result)
	return {
		lines = result.lines or {},
		marks = result.marks or {},
		spans = result.spans or {},
	}
end

--- Encode an IR to a JSON string (the cross-process contract). The producer may
--- be this Lua build or a future Go core; either side decodes the same shape.
---@param result table ProjectionIR
---@return string
function M.encode(result)
	return vim.json.encode(M.to_wire(result))
end

--- Decode a wire IR from JSON.
---@param str string
---@return table { lines, marks, spans }
function M.decode(str)
	return vim.json.decode(str)
end

--- Describe a mark for tests/inspection. Accepts a neutral descriptor (returns
--- it as-is shape) or a legacy dialect mark (classifies it).
---@param mark table
---@return { kind: IRDecorationKind, [string]: any }
function M.classify(mark)
	if mark.kind then
		return {
			kind = mark.kind,
			hl = mark.hl,
			text = mark.text,
			end_col = mark.end_col,
			align = mark.align,
			pos = mark.pos,
			count = mark.lines and #mark.lines or nil,
		}
	end
	return M.to_neutral(mark)
end

return M

File: core/ir/ir.go
// Package ir defines the Projection IR — the editor-agnostic contract emitted by
// the renderer. The Lua plugin produces this JSON today (atlassian.editor.ir);
// this Go package decodes/validates/re-encodes the exact same shape, and is the
// foundation for a future Go core that produces it.
//
// The wire shape (see atlassian/editor/ir.lua):
//
//	{
//	  "lines": ["..."],
//	  "marks": [ { "line", "col", "kind", ...kind-specific } ],
//	  "spans": [ { "line", "col_start", "col_end", "path", "field", "editable" } ]
//	}
//
// Styles are semantic tokens ("heading.2", "strong", "diff.add"), never Neovim
// highlight groups — see atlassian/editor/theme.lua.
package ir

import (
	"encoding/json"
	"fmt"
)

// Chunk is a [text, styleToken] tuple, serialized as a JSON array. The style
// element may be absent (length 1) when the producer attached no style.
type Chunk = []string

// Mark is a neutral decoration descriptor. Kind selects which fields apply:
//
//	highlight    : Hl, EndCol
//	inline_text  : Chunks, Pos
//	eol_text     : Chunks, Align
//	sign         : Text, Hl
//	virt_lines   : Lines, Above
type Mark struct {
	Line int    `json:"line"`
	Col  int    `json:"col"`
	Kind string `json:"kind"`

	Hl     string  `json:"hl,omitempty"`
	Text   string  `json:"text,omitempty"`
	EndCol *int    `json:"end_col,omitempty"`
	Align  string  `json:"align,omitempty"`
	Pos    string  `json:"pos,omitempty"`
	Chunks []Chunk `json:"chunks,omitempty"`
	Lines  []Line  `json:"lines,omitempty"`
	Above  *bool   `json:"above,omitempty"`
}

// Line is one virtual line: a sequence of styled chunks.
type Line = []Chunk

// Span maps a buffer text range back to its ADF node path (for edit sync).
type Span struct {
	Line     int    `json:"line"`
	ColStart int    `json:"col_start"`
	ColEnd   int    `json:"col_end"`
	Path     []int  `json:"path"`
	Field    string `json:"field"`
	Editable bool   `json:"editable"`
}

// ProjectionIR is the wire-serializable build output.
type ProjectionIR struct {
	Lines []string `json:"lines"`
	Marks []Mark   `json:"marks"`
	Spans []Span   `json:"spans"`
}

// Decode parses wire JSON into a ProjectionIR.
func Decode(data []byte) (*ProjectionIR, error) {
	var p ProjectionIR
	if err := json.Unmarshal(data, &p); err != nil {
		return nil, fmt.Errorf("decode IR: %w", err)
	}
	return &p, nil
}

// Encode serializes a ProjectionIR to wire JSON.
func Encode(p *ProjectionIR) ([]byte, error) {
	return json.Marshal(p)
}

// validKinds is the closed set of neutral decoration kinds.
var validKinds = map[string]bool{
	"highlight":   true,
	"inline_text": true,
	"eol_text":    true,
	"virt_lines":  true,
	"sign":        true,
	"unknown":     true,
}

// Validate checks structural invariants of the IR and returns all problems found.
func (p *ProjectionIR) Validate() []error {
	var errs []error
	if p.Lines == nil {
		errs = append(errs, fmt.Errorf("lines is null"))
	}
	for i, m := range p.Marks {
		if !validKinds[m.Kind] {
			errs = append(errs, fmt.Errorf("mark[%d]: unknown kind %q", i, m.Kind))
		}
		if m.Line < 0 || m.Col < 0 {
			errs = append(errs, fmt.Errorf("mark[%d]: negative line/col (%d,%d)", i, m.Line, m.Col))
		}
		if m.Line >= len(p.Lines) {
			errs = append(errs, fmt.Errorf("mark[%d]: line %d out of range (%d lines)", i, m.Line, len(p.Lines)))
		}
	}
	return errs
}

// MarksByKind counts marks grouped by their decoration kind.
func (p *ProjectionIR) MarksByKind() map[string]int {
	out := make(map[string]int)
	for _, m := range p.Marks {
		out[m.Kind]++
	}
	return out
}

// StyleTokens returns the set of distinct style tokens referenced anywhere in
// the IR (highlight/sign hl, plus chunk and virtual-line styles).
func (p *ProjectionIR) StyleTokens() map[string]bool {
	out := make(map[string]bool)
	add := func(s string) {
		if s != "" {
			out[s] = true
		}
	}
	addChunks := func(cs []Chunk) {
		for _, c := range cs {
			if len(c) >= 2 {
				add(c[1])
			}
		}
	}
	for _, m := range p.Marks {
		add(m.Hl)
		addChunks(m.Chunks)
		for _, ln := range m.Lines {
			addChunks(ln)
		}
	}
	return out
}
